src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs

```diff
@@ -32,33 +36,98 @@ public class TestJsonCommand : PSCmdlet
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Prepare a JSON schema.
         /// </summary>
         protected override void BeginProcessing()
         {
-            if (Schema != null)
+            try
             {
-                try
+                if (Schema != null)
                 {
-                    _jschema = JsonSchema.FromJsonAsync(Schema).Result;
+                    try
+                    {
+                        _jschema = JsonSchema.FromJsonAsync(Schema).Result;
+                    }
+                    // Even if only one exception is thrown, it is still wrapped in an AggregateException exception
+                    // https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library
+                    catch (AggregateException ae)
+                    {
+                        ae.Handle(UnwrapException);
+                    }
                 }
-                catch (Exception exc)
+                else if (SchemaFile != null)
                 {
-                    Exception exception = new Exception(TestJsonCmdletStrings.InvalidJsonSchema, exc);
-                    ThrowTerminatingError(new ErrorRecord(exception, "InvalidJsonSchema", ErrorCategory.InvalidData, null));
+                    try
+                    {
+                        string resolvedpath = Context.SessionState.Path.GetUnresolvedProviderPathFromPSPath(SchemaFile);
+                        _jschema = JsonSchema.FromFileAsync(resolvedpath).Result;
+                    }
+                    catch (AggregateException ae)
+                    {
+                        ae.Handle(UnwrapException);
+                    }
                 }
             }
+            // Handle exceptions related to file access to provide more specific error message
+            // https://docs.microsoft.com/en-us/dotnet/standard/io/handling-io-errors
+            catch (Exception e) when (
+                e is IOException ||
+                e is UnauthorizedAccessException ||
+                e is NotSupportedException ||
+                e is SecurityException
+            )
+            {
+                // Do we really need to wrap exception? Not doing this provides more clear error message upfront.
+                // E.g.: "'{}'|Test-Json -SchemaFile c:" results in "Test-Json : Access to the path 'C:\' is denied".
+                Exception exception = new Exception("JSON schema file open failure", e); // TODO: Add resource string
+                ThrowTerminatingError(new ErrorRecord(exception, "JsonSchemaFileOpenFailure", ErrorCategory.OpenError, null));
+            }
+            catch (Exception e)
+            {
+                Exception exception = new Exception(TestJsonCmdletStrings.InvalidJsonSchema, e);
+                ThrowTerminatingError(new ErrorRecord(exception, "InvalidJsonSchema", ErrorCategory.InvalidData, null));
```

Comment:
Please write non-terminating error.

Reply:
The same as above (how `WriteError` is supposed to work here) + there is no chance at all that schema passed via string will ever become valid. Could you please explain in more detail?